cmd/maya-apiserver/cstor-operator/cspc/handler.go

```diff
@@ -219,3 +232,83 @@ func (pc *PoolConfig) createDeployForCSP(csp *apis.NewTestCStorPool) error {
 	}
 	return nil
 }
+
+// getUsedBDCs returns BDCList that is associated with the given cspc
+func (c *Controller) getUsedBDCs(cspc *apis.CStorPoolCluster, namespace string) (*ndmapis.BlockDeviceClaimList, error) {
+	bdcList, err := c.ndmclientset.OpenebsV1alpha1().BlockDeviceClaims(namespace).
+		List(metav1.ListOptions{LabelSelector: string(apis.CStorPoolClusterCPK) + "=" + cspc.Name})
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not list BDCs for CSPC %v", cspc.Name)
+	}
+	return bdcList, nil
+}
+
+// handleDeletion is used to perform operation when the delete timestamp is set
+// on the object
+func (c *Controller) handleDeletion(cspc *apis.CStorPoolCluster, namespace string) error {
+	// get all the BDCs associated with the cspc
+	bdcList, err := c.getUsedBDCs(cspc, namespace)
+	if err != nil {
+		return err
+	}
+
+	// iterate over the bdc and remove the finalizer
+	for _, bdc := range bdcList.Items {
+		err := c.removeBDCFinalizer(&bdc, v1alpha1.CSPCFinalizer)
+		if err != nil {
+			return err
+		}
+	}
+
+	// remove finalizer on cspc
+	err = c.removeCSPCFinalizer(cspc, v1alpha1.CSPCFinalizer)
+	if err != nil {
+		return err
+	}
+
+	// if finalizer is removed successfully, then object will
+	// be removed and no need to reconcile further
+	return nil
+}
+
+// removeBDCFinalizer removes the given finalizer from the BDC
+func (c *Controller) removeBDCFinalizer(bdcObj *ndmapis.BlockDeviceClaim, finalizer string) error {
+	if len(bdcObj.Finalizers) == 0 {
+		return nil
+	}
+
+	bdcObj.Finalizers = util.RemoveString(bdcObj.Finalizers, finalizer)
+
+	// Update is used instead of patch, because when there were 2 finalizers in the object
+	// and tried to remove the first finalizer using patch operation , it was not working. The
+	// patch operation didn't return any error but the object was not getting patched.
+	// using Update() it was possible to remove the finalizer on the BDC
+	_, err := c.ndmclientset.OpenebsV1alpha1().BlockDeviceClaims(bdcObj.Namespace).Update(bdcObj)
+	if err != nil {
+		return errors.Wrapf(err, "failed to remove finalizer from BDC %v", bdcObj.Name)
+	}
+	return nil
+}
+
+// removeCSPCFinalizer will remove finalizer on cspc
+func (c *Controller) removeCSPCFinalizer(cspcObj *apis.CStorPoolCluster, finalizer string) error {
+	if len(cspcObj.Finalizers) == 0 {
+		return nil
+	}
+
+	cspcCopy := cspcObj.DeepCopy()
+	cspcCopy.Finalizers = util.RemoveString(cspcCopy.Finalizers, finalizer)
+	patchBytes, err := strategicmerge.GetPatchData(cspcObj, cspcCopy)
+
+	if err != nil {
```

Comment:
Considering the following case:
- CSPC deletion was initiated. 
- the patch was removed from one of the CSP (say CSP-1). 
- operator is restarted. 
- Deletion is restarted. 
- an attempt is made to remove the finalizer on CSP-1 again. This should pass.

Reply:
CSP does not have a finalizer as of now.